src/shared/settingsConfiguration.ts

```diff
@@ -5,231 +5,590 @@
+ * settings. Use this class when very simple reading/writing to settings is needed.
+ *
+ * When additional logic is needed, prefer {@link fromPackage} for better encapsulation.
  */
-export type SettingsConfiguration = ClassToInterfaceType<DefaultSettingsConfiguration>
-
-export type AwsDevSetting =
-    | 'aws.forceCloud9'
-    | 'aws.dev.forceInstallTools'
-    | 'aws.dev.telemetryEndpoint'
-    | 'aws.dev.telemetryUserPool'
-
-type JSPrimitiveTypeName =
-    | 'undefined'
-    | 'object'
-    | 'boolean'
-    | 'number'
-    | 'bigint'
-    | 'string'
-    | 'symbol'
-    | 'function'
-    | 'object'
-
-export class DefaultSettingsConfiguration implements SettingsConfiguration {
+export class SettingsConfiguration {
     public constructor(
-        private readonly extensionSettingsPrefix: string = 'aws',
-        private readonly log: logger.Logger = logger.getLogger()
+        private readonly target = vscode.ConfigurationTarget.Global,
+        private readonly workspace: Workspace = vscode.workspace
     ) {}
-    public readSetting<T>(settingKey: string): T | undefined
-    public readSetting<T>(settingKey: string, defaultValue: T): T
 
     /**
-     * Reads a vscode setting.
+     * Reads a setting, applying the {@link TypeConstructor} if provided.
      *
-     * @deprecated use getSetting
+     * Note that the absence of a key is indistinguisable from the absence of a value.
+     * That is, `undefined` looks the same across all calls. Any non-existent values are
+     * simply returned as-is without passing through the type cast.
      */
-    public readSetting<T>(settingKey: string, defaultValue?: T): T | undefined {
-        const settings = vscode.workspace.getConfiguration(this.extensionSettingsPrefix)
+    public getSetting(key: string): unknown
+    public getSetting<T>(key: string, type: TypeConstructor<T>): T | undefined
+    public getSetting<T>(key: string, type: TypeConstructor<T>, defaultValue: T): T
+    public getSetting<T>(key: string, type?: TypeConstructor<T>, defaultValue?: T) {
+        const value = this.workspace.getConfiguration().get(key, defaultValue)
 
-        if (!settings) {
-            return defaultValue
+        return !type || value === undefined ? value : cast(value, type)
+    }
+
+    /**
+     * Attempts to write to the settings.
+     *
+     * Settings can only be written to so long as a extension contributes the specified key
+     * in their `package.json`. If no contribution exists, then the write will fail, causing
+     * this method to return false.
+     */
+    public async updateSetting(key: string, value: unknown): Promise<boolean> {
+        try {
+            await this.workspace.getConfiguration().update(key, value, this.target)
+
+            return true
+        } catch (error) {
+            getLogger().warn(`Settings: failed to update "${key}": ${error}`)
+
+            return false
+        }
+    }
+
+    /**
+     * Returns a "slice" of the settings configuration.
```

Comment:
```suggestion
     * Returns a scoped "view" of the settings configuration.
```

Reply:
I think I got too used to the word "view" referring to UI 
Maybe partition is the best word?

Reply:
slice is fine, if view is not as intuitive as I thought. "scoped" as a qualifier makes it clear